swift/ql/src/queries/Security/CWE-079/UnsafeWebViewFetch.ql

```diff
@@ -0,0 +1,143 @@
+      this.asExpr() = call
+    )
+  }
+
+  override string getSourceType() { result = "" }
+}
+
+/**
+ * A sink that is a candidate result for this query, such as certain arguments
+ * to `UIWebView.loadHTMLString`.
+ */
+class Sink extends DataFlow::Node {
+  Expr baseURL;
+
+  Sink() {
+    exists(
+      AbstractFunctionDecl funcDecl, CallExpr call, string funcName, string paramName, int arg,
+      int baseURLarg
+    |
+      // arguments to method calls...
+      exists(string className, ClassDecl c |
+        (
+          // `loadHTMLString`
+          className = ["UIWebView", "WKWebView"] and
+          funcName = "loadHTMLString(_:baseURL:)" and
+          paramName = "string"
+          or
+          // `UIWebView.load`
+          className = "UIWebView" and
+          funcName = "load(_:mimeType:textEncodingName:baseURL:)" and
+          paramName = "data"
+          or
+          // `WKWebView.load`
+          className = "WKWebView" and
+          funcName = "load(_:mimeType:characterEncodingName:baseURL:)" and
+          paramName = "data"
+        ) and
+        c.getName() = className and
+        c.getAMember() = funcDecl and
+        call.getFunction().(ApplyExpr).getStaticTarget() = funcDecl
+      ) and
+      // match up `funcName`, `paramName`, `arg`, `node`.
+      funcDecl.getName() = funcName and
+      funcDecl.getParam(pragma[only_bind_into](arg)).getName() = paramName and
+      call.getArgument(pragma[only_bind_into](arg)).getExpr() = this.asExpr() and
+      // match up `baseURLArg`
+      funcDecl.getParam(pragma[only_bind_into](baseURLarg)).getName() = "baseURL" and
+      call.getArgument(pragma[only_bind_into](baseURLarg)).getExpr() = baseURL
+    )
+  }
+
+  /**
+   * Gets the `baseURL` argument associated with this sink.
+   */
+  Expr getBaseURL() { result = baseURL }
+}
+
+/**
+ * Taint configuration from taint sources to sinks (and `baseURL` arguments)
+ * for this query.
+ */
+class UnsafeWebViewFetchConfig extends TaintTracking::Configuration {
+  UnsafeWebViewFetchConfig() { this = "UnsafeWebViewFetchConfig" }
+
+  override predicate isSource(DataFlow::Node node) { node instanceof RemoteFlowSource }
+
+  override predicate isSink(DataFlow::Node node) {
+    node instanceof Sink or
+    node.asExpr() = any(Sink s).getBaseURL()
+  }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node node1, DataFlow::Node node2) {
+    // allow flow through `try!` and similar constructs
+    // TODO: this should probably be part of DataFlow / TaintTracking.
+    node1.asExpr() = node2.asExpr().(AnyTryExpr).getSubExpr()
+    or
+    // allow flow through `!`
+    // TODO: this should probably be part of DataFlow / TaintTracking.
+    node1.asExpr() = node2.asExpr().(ForceValueExpr).getSubExpr()
+    or
+    // allow flow through string concatenation.
+    // TODO: this should probably be part of TaintTracking.
+    node2.asExpr().(AddExpr).getAnOperand() = node1.asExpr()
```

Comment:
The first two should definitely be taint steps that are included in `TaintTracking`, yeah.

I also think the string concatenation rule should be a taint step in `TaintTracking`. Maybe it should only be a taint step when the other operand is a constant string?

Reply:
I'm actually suspicious flow through `try` and `!` should be *data* flow, never mind taint flow.

Why do you think we might not want taint flow through appending non-constant strings?

Reply:
I just created https://github.com/github/codeql-c-team/issues/1108 to track this by the way, so ideally continue discussion of future taint steps on there.